frontend/src/hooks/auth.ts

```diff
@@ -0,0 +1,97 @@
+import { api } from "@/api";
+import {
+  useMutation,
+  useQueryClient,
+  queryOptions,
+} from "@tanstack/react-query";
+import { enqueueSnackbar } from "notistack";
+import { queryClient } from "@/queryClient";
+import { userProfileQuery } from "./settings";
+import { Routes } from "@/router";
+import { redirect } from "react-router-dom";
+
+export function hasAuthQuery() {
+  // to check if authentication is enabled on the app
+  // used by ProfileDropdown
+  return queryOptions({
+    queryKey: ["hasAuth"],
+    queryFn: async () => await api.hasAuth(),
+    staleTime: Infinity,
+  });
+}
+
+export function isAuthenticatedQuery() {
+  // Query function checks if auth enabled. If it is, calls api.getUserInfo:
+  // ==> If successful, we are logged in (also cache the value of UserInfo)
+  // ==> Otherwise, we are not logged in
+  // Used by Router once the app is loaded to check if we should redirect to "/login"
+  return queryOptions({
+    queryKey: ["isAuthenticated"],
+    queryFn: async () => {
+      try {
+        const hasAuth = await queryClient.fetchQuery(hasAuthQuery());
+        if (!hasAuth) return true;
+        const userInfo = (await api.getUserInfo()).data;
+        queryClient.setQueryData(userProfileQuery().queryKey, userInfo);
+        return true;
+      } catch (error) {
+        return false;
+      }
+    },
+    staleTime: 1000 * 60 * 5, // Expire after 5 minutes
```

Comment:
What exactly is expiring here? The status isAuthenticated? So they'd have to re-login? or?

Why are we calling get user info to check login status? That seems weird. If we have the cookie, we are logged in (presumably). If not, we're not.

If we have a cookie and it's wrong, then someone messed with something manually, so we dc.

Reply:
I removed `staleTime` as it's not needed

> Why are we calling get user info to check login status?
There are two reasons for this:
1. The cookie is created with `httpOnly=True`, meaning we can't access it via javascript. It's optional, but it's more secure that way
2. Even if we have access to the cookie, calling the user info endpoint verifies that the cookie is actually good. Ways the cookie can go bad:
- If someone manually tampers with it, we don't care in this case
- We're planning on doing demos so we're going to change the credentials quite often -> people with the old demo credentials will have a bad cookie
- In the future we will have multiple users feature, changing your password means bad cookie
In all the cases above, if we're just relying on the presence of the cookie, it'll be a bad experience